### PR TITLE
Iox #835 json config reader

### PR DESCRIPTION
**Pull request by grmplmpf**
## Pre-Review Checklist for the PR Author

1. [ ] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [ ] Tests follow the [best practice for testing][testing]
1. [ ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit messages are signed (`git commit -s`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

**Comment by grmplmpf** [comment]
Recreate Pull request to trigger actions

File: iceoryx_posh/cmake/tiny_json/CMakeLists.txt.in [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
```
Comment (sschnitzer) [inline iceoryx_posh/cmake/tiny_json/CMakeLists.txt.in]: Is it correct that some files have Copyright 2021 and other Copyright 2020?

**Comment by shankar-in** [comment]
@oschuene Guess you to want to make this PR to public Iceoryx? Currently this PR is targeting BoschGlobal Iceoryx. It has to point to the master of eclipse iceoryx?

File: iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    iox::cxx::FileReader reader(m_customConfigFilePath, "", iox::cxx::FileReader::ErrorMode::Inform);
+    while (reader.ReadLine(line))
+    {
+        content.append(line);
+    }
+    json_t const* json = json_create(&content[0], mem, NUMBER_OF_JSON_NODES);
+    auto groupOfCurrentProcess = iox::posix::PosixGroup::getGroupOfCurrentProcess().getName();
+
+    json_t const* general;
+    general = json_getProperty(json, "general");
```
Comment (sschnitzer) [inline iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp:136]: If I got the tiny_json API correctly, json_create() will return a nullptr if NUMBER_OF_JSON_NODES is too small. And json_getProperty does not check on nullptr, so potentially, this dereferences a nullpointer.

File: iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        return iox::cxx::success<iox::RouDiConfig_t>(defaultConfig);
+    }
+
+    iox::cxx::FileReader reader(m_customConfigFilePath, "", iox::cxx::FileReader::ErrorMode::Inform);
+    while (reader.ReadLine(line))
+    {
+        content.append(line);
+    }
+    json_t const* json = json_create(&content[0], mem, NUMBER_OF_JSON_NODES);
+    auto groupOfCurrentProcess = iox::posix::PosixGroup::getGroupOfCurrentProcess().getName();
```
Comment (sschnitzer) [inline iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp]: Please place this line right before the for-loop it is used in. Improves readability.

File: iceoryx_posh/include/iceoryx_posh/roudi/roudi_config_json_file_provider.hpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+{
+namespace roudi
+{
+static constexpr char defaultConfigJsonPath[] = "/etc/iceoryx/roudi_config.json";
+
+/// @brief Reader for Roudi Configuration using json config file
+class JsonRouDiConfigFileProvider : public iox::roudi::RouDiConfigFileProvider
+{
+  private:
+    constexpr static uint32_t NUMBER_OF_JSON_NODES = 1024;
```
Comment (sschnitzer) [inline iceoryx_posh/include/iceoryx_posh/roudi/roudi_config_json_file_provider.hpp]: This value seems to be somehow arbitrarily chosen. The TOML implementation does not have such a limit. If you really see the need to limit the stack memory usage, then please put this value into `iceoryx_posh_types.hpp ` and document it properly. If not (which is what I assume), just calculate the value using MAX_NUMBER_OF_MEMPOOLS and MAX_SHM_SEGMENTS, to make sure we neither prevent huge config files nor waste stack memory by using too high values.

File: iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+                 "\t\t\t\t\t\"size\":10000\n"
+                 "\t\t\t\t\t\"count\":\"\"\n"
+                 "\t\t\t\t}\n"
+                 "\t\t\t]\n"
+                 "\t\t}\n"
+                 "\t]\n"
+                 "}\n";
+    ASSERT_EQ(iox::roudi::RouDiConfigFileParseError::MEMPOOL_WITHOUT_CHUNK_COUNT, parseJson(str).get_error());
+}
+
```
Comment (sschnitzer) [inline iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp:441]: A test that tries to exceed NUMBER_OF_JSON_NODES is missing.

**Review by sschnitzer — CHANGES_REQUESTED** [review]
See findings.

File: iceoryx_posh/cmake/tiny_json/CMakeLists.txt.in (hunk 1 above)
Comment (grmplmpf) [reply iceoryx_posh/cmake/tiny_json/CMakeLists.txt.in]: Fixed all copyright years to 2021

File: iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp (hunk 2 above)
Comment (grmplmpf) [reply iceoryx_posh/source/roudi/roudi_config_json_file_provider.cpp:136]: Added nullptr check

File: iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp (hunk 5 above)
Comment (grmplmpf) [reply iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp:441]: added

**Pull request closed without merging**